File: src/CSharp/117.PefectBinTreeConnectRightII.cs

using System;
using System.Linq;
using System.Text;
using System.Collections.Generic;


namespace PeefyLeetCode.PefectBinTreeConnectRightII
{

public class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}

    public Node(int _val,Node _left,Node _right,Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};

class Solution {
    public Node Connect(Node root) {
        if (root == null)
            return null;
        Node dummy = new Node(-1, null, null, null);
        Node pre = dummy;
        Node head = root;
        while (root != null) {
            if (root.left != null) {
                pre.next = root.left;
                pre = pre.next;
            }
            if (root.right != null) {
                pre.next = root.right;
                pre = pre.next;
            }
            root = root.next;
            if (root == null) {
                pre = dummy;
                root = dummy.next;
                dummy.next = null;
            }
        }
        return head;
    }
}
}

File: src/Python/117.PefectBinTreeConnectRightII.py

class Node:
    def __init__(self, val, left = None, right = None, next = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next


class Solution:
    def connect(self, root):
        if root is None:
            return root
        stack = [root]
        while len(stack) > 0:
            tmp = []
            lastnode = None
            for i in range(len(stack)):
                node = stack[i]
                node.next = lastnode
                lastnode = stack[i]
                if node.right is not None:
                    tmp.append(node.right)
                if node.left is not None:
                    tmp.append(node.left)
            stack = tmp
        return root


if __name__ == "__main__":
    solution = Solution()
    node = Node(1)
    node.left = Node(2)
    node.right = Node(3)
    node.left.left = Node(4)
    node.left.right = Node(5)
    node.right.left = Node(6)
    node.right.right = Node(7)
    print(solution.connect(node))
    i = 1